File: solutions/elixir/resistor-color-trio/5/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @doc """
  Calculate the resistance value in ohms from resistor colors
  """
  
  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label(colors) do
    colors = Enum.take(colors, 3)
    {value_1, _} = color_value(List.first(colors))
    {value_2, _} = color_value(List.first(tl(colors)))
    {_, zeroes} = color_value(List.last(colors))

    number_value = List.to_integer(~c"#{value_1}#{value_2}#{zeroes}")
    length_of_num = String.length("#{number_value}")

    {val, unit} = case length_of_num do
      length_of_num when length_of_num in 1..3 -> {number_value, :ohms}
      length_of_num when length_of_num in 4..6 -> {div(number_value, 1000), :kiloohms}
      length_of_num when length_of_num in 7..9 -> {div(number_value, 1_000_000), :megaohms}
      length_of_num when length_of_num >= 10 -> {div(number_value, 1_000_000_000), :gigaohms}
    end
    
    {val, unit}
  end

  defp color_value(color) do
    case color do
      :black -> {0, ""}
      :brown -> {1, "0"}
      :red -> {2, "00"}
      :orange -> {3, "000"}
      :yellow -> {4, "0000"}
      :green -> {5, "00000"}
      :blue -> {6, "000000"}
      :violet -> {7, "0000000"}
      :grey ->{ 8, "00000000"}
      :white -> {9, "000000000"}
    end
  end
  
end

File: solutions/c/resistor-color-trio/2/resistor_color_trio.c
#include "resistor_color_trio.h"
#include <stdint.h>

static uint16_t band_to_int(resistor_band_t b) {
  uint16_t result;

  switch (b) {
  case BLACK:  result = 0; break;
  case BROWN:  result = 1; break;
  case RED:    result = 2; break;
  case ORANGE: result = 3; break;
  case YELLOW: result = 4; break;
  case GREEN:  result = 5; break;
  case BLUE:   result = 6; break;
  case VIOLET: result = 7; break;
  case GREY:   result = 8; break;
  case WHITE:  result = 9; break;
  }
  return result;
}


resistor_value_t color_code(resistor_band_t band[]) {
  uint64_t ohms = 10 * band_to_int(band[0]) + band_to_int(band[1]);
  resistor_value_t final;

  int power = band_to_int(band[2]);
  for (int j = 0; j < power; j++) {
    ohms *= 10;
  }

  if (ohms % 1000000000 == 0 && ohms > 0) {
    final.value = ohms / 1000000000;
    final.unit  = GIGAOHMS;
  } else if (ohms % 1000000 == 0 && ohms > 0) {
    final.value = ohms / 1000000;
    final.unit  = MEGAOHMS;
  } else if (ohms % 1000 == 0 && ohms > 0) {
    final.value = ohms / 1000;
    final.unit  = KILOOHMS;
  } else {
    final.value = ohms;
    final.unit  = OHMS;
  }

  return final;
}
